cmd/qemuconf.go

```diff
@@ -0,0 +1,332 @@
+package cmd
+
+import (
+	"fmt"
+	"github.com/lf-edge/eden/pkg/utils"
+	"github.com/spf13/cobra"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+const (
+	defaultQemuFileToSave = "qemu.conf"
+	defaultQemuCpus       = 4
+	defaultQemuMemory     = 4096
+)
+
+var (
+	qemuFileToSave   string
+	qemuCpus         int
+	qemuMemory       int
+	qemuFirmwarePath string
+	qemuConfigPath   string
+	qemuImagePath    string
+	qemuDTBPath      string
+	qemuHostFwd      = map[string]string{"2222": "22"}
+	qemuSocketPath   string
+)
+
+type qemuDrive struct {
+	file      string
+	format    string
+	directory bool
+	label     string
+}
+
+func (drive *qemuDrive) toString() string {
+	output := fmt.Sprintln("[drive]")
```

Comment:
Building qemu configuration like this bit by bit is doable, but brittle. What we should do instead is a templating approach of some kind. Here's how we do it in EVE itself: https://github.com/lf-edge/eve/blob/master/pkg/pillar/hypervisor/kvm.go#L40

And then https://github.com/lf-edge/eve/blob/master/pkg/pillar/hypervisor/kvm.go#L335

Please consider this approach instead.

Now, I'm actually not even saying it should be done based on Golang templates, given that it is highly likely that a user of Adam may want to tweak this config -- perhaps we can keep it externally as a template and then  decide whether we use golang templates or something like Mustache for actual templating.

@deitch may have other ideas here.

In fact, @deitch -- I am thinking that this could be useful in linuxkit itself -- basically providing an override qemu config for `linuxkit run qemu` use case -- what do you think?

Reply:
For the first part, I get the appeal of doing it in go inside eden. Eden is a single command, and so being able to say, "give me qemu config" makes sense. But doing it bit by bit as rough, unless you are willing to write software that can manage the entire qemu config. However, to be fair, linuxkit does _precisely_ that, see [here](https://github.com/linuxkit/linuxkit/blob/master/src/cmd/linuxkit/run_qemu.go#L315-L337).

Is there any reason not to use all of those run options, or at least the struct? For that matter, we could externalize the private parts from linuxkit as a library so it can be leveraged elsewhere.

If that is what you want, raise an issue there.

Reply:
I'd like to have the qemu configuration template that can be externalized just the same way we do it in EVE @deitch -- it is much easier to do small tweak that way to how qemu is run, as opposed to tweaking a ginourmous command line.

Also, I honestly really think that lk should have an option of consuming a fully formed qemu configuration file. May be ON TOP of what it does today -- but it would be very helpful.